File: Data Structures & Algorithms/contains-duplicate-ii/submission-5.rs
use std::collections::HashSet;

impl Solution {
    pub fn contains_nearby_duplicate(nums: Vec<i32>, k: i32) -> bool {
        //initialize our window HashSet so that we can add/remove elements
        let mut window: HashSet<i32> = HashSet::new();

        let mut l: usize = 0;

        for r in 0..nums.len() 
        {
            //if our window length > k, move l right to shrink
            if (r - l) as i32 > k 
            {
                window.remove(&nums[l]);
                l += 1;
            }

            //if we find number in our hashset, return true
            if window.contains(&nums[r]) 
            {
                return true;
            }

            //add our current element to hash set
            window.insert(nums[r]);
        }

        false
    }
}

File: Data Structures & Algorithms/search-for-word/submission-5.py
"""
I think we can try a basic backtracking solution where we call our backtracking
function on each cell in the board and move up/down/left/right until we find
the element we're looking for or hit a base case of going out of bounds or
hitting a cell we've already seen or a cell that has a word that doesn't
make sense.
"""

class Solution:
    def __init__(self):
        self.seen = set([])
        self.n = -1
        self.m = -1
    def backtrack(self, board, word, row, col):
        if not word:
            return True #base case for empty word

        #check for bounds base case
        if row < 0 or row >= self.n or col < 0 or col >= self.m:
            return False
        
        #check for seen in current word base case
        if (row, col) in self.seen:
            return False
        
        #check if first character valid
        if board[row][col] != word[0]:
            return False
        
        if len(word) == 1:
            return True #we found last character to be True
            # and since the last character matches we can stop
    
        self.seen.add((row, col))

        ret = self.backtrack(board, word[1:], row - 1, col) or self.backtrack(board, word[1:], row + 1, col) or self.backtrack(board, word[1:], row, col - 1) or self.backtrack(board, word[1:], row, col + 1)

        #dont forget to actually backtrack here!
        self.seen.remove((row, col))
        
        return ret
    def exist(self, board: List[List[str]], word: str) -> bool:
        self.n = len(board)
        self.m = len(board[0])

        for r in range(len(board)):
            for c in range(len(board[0])):
                self.seen = set([])#reset
                if self.backtrack(board, word, r, c):
                    return True
        return False 
        #if we get to return above it means we definitely wont find
        #word